File: tests/test_adversarial_coverage_challenger2.mjs
import test from "node:test";
import assert from "node:assert/strict";
import { FleetState, extractValidTailscaleIp, normalizeDeviceId } from "../worker/fleet_state.js";

class MockStorage {
  constructor() {
    this.store = new Map();
  }
  async get(key) {
    const val = this.store.get(key);
    return val !== undefined ? JSON.parse(JSON.stringify(val)) : undefined;
  }
  async put(key, value) {
    this.store.set(key, JSON.parse(JSON.stringify(value)));
  }
  async delete(key) {
    this.store.delete(key);
  }
}

function createMockFleetState(telegramCapture) {
  const env = {
    TEST_ENV: true,
    FLEET_SHARED_SECRET: "test-secret",
    TELEGRAM_BOT_TOKEN: "mock-token-123",
    TELEGRAM_ADMIN_USER_ID: "12345678"
  };

  const storage = new MockStorage();
  const ctx = {
    storage,
    sockets: new Map(),
    getWebSockets() { return []; }
  };

  return new FleetState(ctx, env);
}

test("Challenger Iter2-2 Adversarial FleetState & Tailscale Test Suite", async (t) => {
  let telegramCalls = [];
  const originalFetch = globalThis.fetch;
  globalThis.fetch = async (url, options = {}) => {
    if (typeof url === "string" && url.includes("api.telegram.org")) {
      const body = options.body ? JSON.parse(options.body) : {};
      telegramCalls.push({ url, body });
      return { ok: true, json: async () => ({ ok: true }) };
    }
    return originalFetch(url, options);
  };

  t.after(() => {
    globalThis.fetch = originalFetch;
  });

  // =========================================================================
  // 1. CONCURRENCY: 20 Devices Parallel Queue & Concurrent ACK Dispatch
  // =========================================================================
  await t.test("Concurrency: 20 devices parallel queue and concurrent ACK dispatch", async () => {
    telegramCalls = [];
    const fleet = createMockFleetState();

    const deviceIds = Array.from({ length: 20 }, (_, i) => `m${i + 1}`);

    // Register all 20 devices as online
    for (const devId of deviceIds) {
      await fleet.handleHeartbeat(new Request("https://localhost/report", {
        method: "POST",
        headers: { "x-device-secret": "test-secret", "content-type": "application/json" },
        body: JSON.stringify({
          device_id: devId,
          device_group: "NOVA",
          version: "phanserver-delta-agent-1.0.0",
          capabilities: ["allocate_server_2pc", "update_delta", "control_tailscale"]
        })
      }));
    }

    // Queue Tailscale ON for all 20 devices in a single batch
    const queueRes = await (await fleet.controlFleetHub(new Request("https://localhost/aot/hub/control", {
      method: "POST",
      body: JSON.stringify({
        protocol: "fleet-batch-v1",
        kind: "control_tailscale",
        mode: "on",
        target_device_ids: deviceIds,
        telegram_chat_id: 12345678
      })
    }))).json();

    assert.equal(queueRes.ok, true);
    const actionId = queueRes.tailscale.action_id;
    assert.equal(queueRes.tailscale.devices.length, 20);

    // Concurrently dispatch ACKs for all 20 devices with distinct valid CGNAT IPs
    const ackPromises = deviceIds.map((devId, idx) => {
      const ip = `100.80.${idx + 1}.55`;
      return fleet.dispatchFleetAck(new Request("https://localhost/aot/ack", {
        method: "POST",
        headers: { "x-device-secret": "test-secret", "content-type": "application/json" },
        body: JSON.stringify({
          protocol: "fleet-batch-v1",
          batch_action: "CONTROL_TAILSCALE",
          device_id: devId,
          action_id: actionId,
          status: "OPENED",
          executed: true,
          details: `CONNECTED: ${ip}`
        })
      })).then((r) => r.json());
    });

    const ackResults = await Promise.all(ackPromises);
    assert.equal(ackResults.length, 20);
    for (const res of ackResults) {
      assert.equal(res.ok, true);
      assert.equal(res.status, "OPENED");
    }

    // Verify all 20 Telegram messages were sent
    assert.equal(telegramCalls.length, 20);
    for (let i = 0; i < 20; i++) {
      const call = telegramCalls[i];
      assert.match(call.body.text, /ĐÃ BẬT TAILSCALE THÀNH CÔNG!/);
      assert.match(call.body.text, /100\.80\.\d+\.55/);
    }
  });

  // =========================================================================
  // 2. IDEMPOTENCY: Duplicate ACK Replay Stress
  // =========================================================================
  await t.test("Idempotency: Replaying identical ACK does not corrupt state or fail", async () => {
    telegramCalls = [];
    const fleet = createMockFleetState();
    const devId = "m77";

    await fleet.handleHeartbeat(new Request("https://localhost/report", {
      method: "POST",
      headers: { "x-device-secret": "test-secret", "content-type": "application/json" },
      body: JSON.stringify({
        device_id: devId,
        device_group: "NOVA",
        capabilities: ["allocate_server_2pc", "update_delta", "control_tailscale"]
      })
    }));

    const queueRes = await (await fleet.controlFleetHub(new Request("https://localhost/aot/hub/control", {
      method: "POST",
      body: JSON.stringify({
        protocol: "fleet-batch-v1",
        kind: "control_tailscale",
        mode: "on",
        target_device_ids: [devId],
        telegram_chat_id: 12345678
      })
    }))).json();

    const actionId = queueRes.tailscale.action_id;

    // Send ACK first time
    const ack1 = await (await fleet.dispatchFleetAck(new Request("https://localhost/aot/ack", {
      method: "POST",
      headers: { "x-device-secret": "test-secret", "content-type": "application/json" },
      body: JSON.stringify({
        protocol: "fleet-batch-v1",
        batch_action: "CONTROL_TAILSCALE",
        device_id: devId,
        action_id: actionId,
        status: "OPENED",
        executed: true,
        details: "CONNECTED: 100.80.175.55"
      })
    }))).json();
    assert.equal(ack1.ok, true);
    assert.equal(ack1.status, "OPENED");

    // Send identical ACK 5 more times
    for (let i = 0; i < 5; i++) {
      const ackN = await (await fleet.dispatchFleetAck(new Request("https://localhost/aot/ack", {
        method: "POST",
        headers: { "x-device-secret": "test-secret", "content-type": "application/json" },
        body: JSON.stringify({
          protocol: "fleet-batch-v1",
          batch_action: "CONTROL_TAILSCALE",
          device_id: devId,
          action_id: actionId,
          status: "OPENED",
          executed: true,
          details: "CONNECTED: 100.80.175.55"
        })
      }))).json();
      assert.equal(ackN.ok, true);
      assert.equal(ackN.status, "OPENED");
    }
  });

  // =========================================================================
  // 3. RAPID MODE SWITCHING & STATUS INQUIRY
  // =========================================================================
  await t.test("Rapid Mode Switching: ON -> STATUS (connected) -> OFF -> STATUS (disconnected)", async () => {
    telegramCalls = [];
    const fleet = createMockFleetState();
    const devId = "m77";

    await fleet.handleHeartbeat(new Request("https://localhost/report", {
      method: "POST",
      headers: { "x-device-secret": "test-secret", "content-type": "application/json" },
      body: JSON.stringify({
        device_id: devId,
        device_group: "NOVA",
        capabilities: ["allocate_server_2pc", "update_delta", "control_tailscale"]
      })
    }));

    // Step A: Turn ON
    const qOn = await (await fleet.controlFleetHub(new Request("https://localhost/aot/hub/control", {
      method: "POST",
      body: JSON.stringify({
        protocol: "fleet-batch-v1",
        kind: "control_tailscale",
        mode: "on",
        target_device_ids: [devId],
        telegram_chat_id: 12345678
      })
    }))).json();
    telegramCalls = [];
    await fleet.dispatchFleetAck(new Request("https://localhost/aot/ack", {
      method: "POST",
      body: JSON.stringify({
        protocol: "fleet-batch-v1",
        batch_action: "CONTROL_TAILSCALE",
        device_id: devId,
        action_id: qOn.tailscale.action_id,
        status: "OPENED",
        executed: true,
        details: "CONNECTED: 100.80.175.55"
      })
    }));
    assert.equal(telegramCalls.length, 1);
    assert.match(telegramCalls[0].body.text, /ĐÃ BẬT TAILSCALE THÀNH CÔNG!/);
    assert.match(telegramCalls[0].body.text, /100\.80\.175\.55/);

    // Step B: Query STATUS while connected
    const qStatus1 = await (await fleet.controlFleetHub(new Request("https://localhost/aot/hub/control", {
      method: "POST",
      body: JSON.stringify({
        protocol: "fleet-batch-v1",
        kind: "control_tailscale",
        mode: "status",
        target_device_ids: [devId],
        telegram_chat_id: 12345678
      })
    }))).json();
    telegramCalls = [];
    await fleet.dispatchFleetAck(new Request("https://localhost/aot/ack", {
      method: "POST",
      body: JSON.stringify({
        protocol: "fleet-batch-v1",
        batch_action: "CONTROL_TAILSCALE",
        device_id: devId,
        action_id: qStatus1.tailscale.action_id,
        status: "OPENED",
        executed: true,
        details: "CONNECTED: 100.80.175.55"
      })
    }));
    assert.equal(telegramCalls.length, 1);
    assert.match(telegramCalls[0].body.text, /TRẠNG THÁI TAILSCALE: CONNECTED \(100\.80\.175\.55\)/);

    // Step C: Turn OFF
    const qOff = await (await fleet.controlFleetHub(new Request("https://localhost/aot/hub/control", {
      method: "POST",
      body: JSON.stringify({
        protocol: "fleet-batch-v1",
        kind: "control_tailscale",
        mode: "off",
        target_device_ids: [devId],
        telegram_chat_id: 12345678
      })
    }))).json();
    telegramCalls = [];
    await fleet.dispatchFleetAck(new Request("https://localhost/aot/ack", {
      method: "POST",
      body: JSON.stringify({
        protocol: "fleet-batch-v1",
        batch_action: "CONTROL_TAILSCALE",
        device_id: devId,
        action_id: qOff.tailscale.action_id,
        status: "OPENED",
        executed: true,
        details: "DISCONNECTED"
      })
    }));
    assert.equal(telegramCalls.length, 1);
    assert.match(telegramCalls[0].body.text, /ĐÃ TẮT TAILSCALE THÀNH CÔNG!/);

    // Step D: Query STATUS while disconnected
    const qStatus2 = await (await fleet.controlFleetHub(new Request("https://localhost/aot/hub/control", {
      method: "POST",
      body: JSON.stringify({
        protocol: "fleet-batch-v1",
        kind: "control_tailscale",
        mode: "status",
        target_device_ids: [devId],
        telegram_chat_id: 12345678
      })
    }))).json();
    telegramCalls = [];
    await fleet.dispatchFleetAck(new Request("https://localhost/aot/ack", {
      method: "POST",
      body: JSON.stringify({
        protocol: "fleet-batch-v1",
        batch_action: "CONTROL_TAILSCALE",
        device_id: devId,
        action_id: qStatus2.tailscale.action_id,
        status: "OPENED",
        executed: true,
        details: "DISCONNECTED"
      })
    }));
    assert.equal(telegramCalls.length, 1);
    assert.match(telegramCalls[0].body.text, /TRẠNG THÁI TAILSCALE: DISCONNECTED/);
  });

  // =========================================================================
  // 4. DEEP IP REGEX & EXTRACTION FUZZING (extractValidTailscaleIp)
  // =========================================================================
  await t.test("IP Extraction Fuzzing: exhaustive boundary test matrix", async () => {
    const validPairs = [
      ["100.80.175.55", "100.80.175.55"],
      ["CONNECTED: 100.80.175.55", "100.80.175.55"],
      ["IP is 100.64.0.1 on tun0", "100.64.0.1"],
      ["100.0.0.0", "100.0.0.0"],
      ["100.255.255.255", "100.255.255.255"],
      ["[100.115.92.14]", "100.115.92.14"],
      ["(100.115.92.14)", "100.115.92.14"],
    ];
    for (const [input, expected] of validPairs) {
      assert.equal(extractValidTailscaleIp(input), expected, `Failed for input: ${input}`);
    }

    const invalidInputs = [
      "100.300.1.1",            // octet > 255
      "100.1.256.1",
      "100.1.2.256",
      "100.1.2.2555",           // 4-digit octet suffix
      "1100.1.2.3",             // 4-digit prefix
      "100.1.2.3.4",            // 5 octets
      ".100.1.2.3",             // leading dot
      "100.1.2.3:80",           // port
      "100.1.2.3/24",           // CIDR
      "192.168.1.1",            // non-100 prefix
      "10.0.0.1",
      "172.16.0.1",
      "127.0.0.1",
      "TRIGGERED",
      "DISCONNECTED",
      "",
      null,
      undefined
    ];
    for (const input of invalidInputs) {
      assert.equal(extractValidTailscaleIp(input), null, `Expected null for: ${input}`);
    }
  });

  // =========================================================================
  // 5. HTML ESCAPING IN FAILURE REASONS
  // =========================================================================
  await t.test("HTML escaping in error reasons prevents injection into Telegram", async () => {
    telegramCalls = [];
    const fleet = createMockFleetState();
    const devId = "m77";

    await fleet.handleHeartbeat(new Request("https://localhost/report", {
      method: "POST",
      headers: { "x-device-secret": "test-secret", "content-type": "application/json" },
      body: JSON.stringify({
        device_id: devId,
        device_group: "NOVA",
        capabilities: ["allocate_server_2pc", "update_delta", "control_tailscale"]
      })
    }));

    const q = await (await fleet.controlFleetHub(new Request("https://localhost/aot/hub/control", {
      method: "POST",
      body: JSON.stringify({
        protocol: "fleet-batch-v1",
        kind: "control_tailscale",
        mode: "on",
        target_device_ids: [devId],
        telegram_chat_id: 12345678
      })
    }))).json();

    await fleet.dispatchFleetAck(new Request("https://localhost/aot/ack", {
      method: "POST",
      body: JSON.stringify({
        protocol: "fleet-batch-v1",
        batch_action: "CONTROL_TAILSCALE",
        device_id: devId,
        action_id: q.tailscale.action_id,
        status: "FAILED",
        executed: false,
        reason: "<script>alert('pwn')</script> & error <b>bold</b>"
      })
    }));

    assert.equal(telegramCalls.length, 1);
    const text = telegramCalls[0].body.text;
    assert.match(text, /&lt;script&gt;alert\('pwn'\)&lt;\/script&gt;/);
    assert.match(text, /&amp; error/);
    assert.match(text, /&lt;b&gt;bold&lt;\/b&gt;/);
  });
});

File: tests/test_adversarial_coverage_challenger2.py
#!/usr/bin/env python3
"""Challenger Iter2-2 Adversarial Coverage & Stress Test Harness (Python).
Tests:
- Screen orientation, coordinates fuzzing, extreme aspect ratios, shell fallback
- Idempotency & state caching persistence under re-entrant executions
- Tailscale IP validation & boundary regex lookarounds
- Subprocess error handling & signal termination
- R4 Compliance verification (zero real UgPhone commands or leaks)
"""

import json
import os
import pathlib
import re
import subprocess
import sys
import tempfile
import unittest
from unittest import mock

PROJECT_ROOT = pathlib.Path(__file__).resolve().parent.parent
if str(PROJECT_ROOT) not in sys.path:
    sys.path.insert(0, str(PROJECT_ROOT))

from agent.agent import (
    build_tailscale_command,
    compute_screen_coordinates,
    handle_incoming_batch_action,
    validate_tailscale_cgnat_ip,
)


class TestOrientationAdversarialStress(unittest.TestCase):
    """Stress testing screen orientation, resolution fuzzing, and coordinate safety."""

    def test_orientation_fuzzing_resolutions(self):
        """Fuzz a wide matrix of aspect ratios and rotations."""
        test_dimensions = [
            (720, 1280),    # Standard 720p portrait
            (1280, 720),    # Standard 720p landscape
            (1080, 1920),   # Standard 1080p
            (1920, 1080),
            (1080, 2400),   # 20:9
            (1440, 3120),   # 19.5:9
            (1440, 2560),   # 16:9 2K
            (720, 720),     # Square 1:1
            (1080, 1080),   # Square 1:1
            (480, 800),     # WVGA
            (320, 480),     # HVGA legacy
            (2160, 3840),   # 4K portrait
            (3840, 2160),   # 4K landscape
        ]
        rotations = [0, 1, 2, 3, 4, -1, "0", "1"]

        for w, h in test_dimensions:
            for rot in rotations:
                coords = compute_screen_coordinates(w, h, rotation=rot)

                self.assertIn("is_landscape", coords)
                self.assertIn("width", coords)
                self.assertIn("height", coords)
                self.assertIn("toggle_x", coords)
                self.assertIn("toggle_y", coords)
                self.assertIn("center_x", coords)
                self.assertIn("center_y", coords)

                actual_w = coords["width"]
                actual_h = coords["height"]

                # Ensure dimensions match max/min based on landscape status
                if coords["is_landscape"]:
                    self.assertEqual(actual_w, max(int(w), int(h)))
                    self.assertEqual(actual_h, min(int(w), int(h)))
                else:
                    self.assertEqual(actual_w, min(int(w), int(h)))
                    self.assertEqual(actual_h, max(int(w), int(h)))

                # Coordinates must be strictly within screen boundary
                self.assertGreater(coords["toggle_x"], 0)
                self.assertLess(coords["toggle_x"], actual_w)
                self.assertGreater(coords["toggle_y"], 0)
                self.assertLess(coords["toggle_y"], actual_h)
                self.assertGreater(coords["center_x"], 0)
                self.assertLess(coords["center_x"], actual_w)
                self.assertGreater(coords["center_y"], 0)
                self.assertLess(coords["center_y"], actual_h)

    def test_shell_script_rotation_detection_fallback(self):
        """Verify the shell rotation fallback logic without hardware."""
        # Shell logic extracted from build_tailscale_command:
        test_shell_snippet = """
ROTATION=""
[ -z "$ROTATION" ] && ROTATION=0
RAW_SIZE="720x1280"
DIM_W=$(echo "$RAW_SIZE" | cut -d'x' -f1)
DIM_H=$(echo "$RAW_SIZE" | cut -d'x' -f2)
if [ -n "$DIM_W" ] && [ -n "$DIM_H" ] && [ "$DIM_W" -gt 0 ] 2>/dev/null; then
    if [ "$DIM_W" -gt "$DIM_H" ]; then
        MAX_D="$DIM_W"
        MIN_D="$DIM_H"
    else
        MAX_D="$DIM_H"
        MIN_D="$DIM_W"
    fi
else
    MIN_D=720
    MAX_D=1280
fi
if [ "$ROTATION" -eq 1 ] || [ "$ROTATION" -eq 3 ]; then
    WIDTH="$MAX_D"
    HEIGHT="$MIN_D"
    TOGGLE_X=$((WIDTH * 92 / 100))
    TOGGLE_Y=$((HEIGHT * 12 / 100))
else
    WIDTH="$MIN_D"
    HEIGHT="$MAX_D"
    TOGGLE_X=$((WIDTH * 88 / 100))
    TOGGLE_Y=$((HEIGHT * 8 / 100))
fi
echo "$WIDTH $HEIGHT $TOGGLE_X $TOGGLE_Y"
"""
        res = subprocess.run(["sh", "-c", test_shell_snippet], capture_output=True, text=True)
        self.assertEqual(res.returncode, 0)
        out = res.stdout.strip().split()
        self.assertEqual(out, ["720", "1280", "633", "102"])


class TestAgentIdempotencyAdversarialStress(unittest.TestCase):
    """Stress testing agent action caching, repeated calls, and persistent state reload."""

    def setUp(self):
        self.temp_dir = tempfile.TemporaryDirectory()
        self.root_path = pathlib.Path(self.temp_dir.name)
        self.state_path = self.root_path / "state.json"
        self.links_path = self.root_path / "server_links.txt"
        self.report_url = "https://mock.worker/report"
        self.secret = "mock-secret"
        self.device_id = "m77"

    def tearDown(self):
        self.temp_dir.cleanup()

    @mock.patch("agent.agent.send_ack", return_value=True)
    @mock.patch("agent.agent.subprocess.run")
    def test_repeated_action_id_executes_subprocess_exactly_once(self, mock_subproc, mock_ack):
        """Calling handle_incoming_batch_action 20 times with same action_id must invoke subprocess once."""
        mock_subproc.return_value.returncode = 0
        mock_subproc.return_value.stdout = "CONNECTED: 100.80.175.55"
        mock_subproc.return_value.stderr = ""

        state = {}
        action_id = "stress-idempotency-001"
        msg = {
            "protocol": "fleet-batch-v1",
            "action": "CONTROL_TAILSCALE",
            "action_id": action_id,
            "mode": "on",
            "target_device_ids": [self.device_id],
        }

        # 20 sequential calls
        for i in range(20):
            res = handle_incoming_batch_action(
                msg, self.device_id, self.report_url, self.secret, state, self.state_path, self.links_path
            )
            self.assertTrue(res)

        self.assertEqual(mock_subproc.call_count, 1, "Subprocess was invoked more than once!")
        self.assertEqual(mock_ack.call_count, 20, "send_ack should be sent on each replay")
        for call in mock_ack.call_args_list:
            self.assertEqual(call.kwargs["status"], "OPENED")
            self.assertEqual(call.kwargs["details"], "CONNECTED: 100.80.175.55")

        # Verify state file was written and survives fresh reload
        self.assertTrue(self.state_path.exists())
        saved_state = json.loads(self.state_path.read_text(encoding="utf-8"))
        self.assertIn("tailscale_action_results", saved_state)
        self.assertIn(action_id, saved_state["tailscale_action_results"])
        cached = saved_state["tailscale_action_results"][action_id]
        self.assertEqual(cached["status"], "OPENED")
        self.assertEqual(cached["details"], "CONNECTED: 100.80.175.55")

        # Fresh agent process simulation: load state from file and call again
        fresh_state = json.loads(self.state_path.read_text(encoding="utf-8"))
        mock_ack.reset_mock()
        handle_incoming_batch_action(
            msg, self.device_id, self.report_url, self.secret, fresh_state, self.state_path, self.links_path
        )
        self.assertEqual(mock_subproc.call_count, 1, "Subprocess invoked even with loaded persistent state!")
        self.assertEqual(mock_ack.call_args.kwargs["status"], "OPENED")

    @mock.patch("agent.agent.send_ack", return_value=True)
    @mock.patch("agent.agent.subprocess.run")
    def test_rapid_alternating_modes_distinct_action_ids(self, mock_subproc, mock_ack):
        """Simulate rapid alternating commands (on -> off -> status -> on -> off)."""
        state = {}
        modes_and_outputs = [
            ("on", "CONNECTED: 100.80.175.55", 0, "OPENED", "CONNECTED: 100.80.175.55"),
            ("off", "DISCONNECTED", 0, "OPENED", "DISCONNECTED"),
            ("status", "DISCONNECTED", 0, "OPENED", "DISCONNECTED"),
            ("on", "CONNECTED: 100.115.92.14", 0, "OPENED", "CONNECTED: 100.115.92.14"),
            ("status", "CONNECTED: 100.115.92.14", 0, "OPENED", "CONNECTED: 100.115.92.14"),
            ("off", "DISCONNECTED", 0, "OPENED", "DISCONNECTED"),
        ]

        for idx, (mode, stdout_val, code, exp_status, exp_details) in enumerate(modes_and_outputs):
            mock_subproc.return_value.returncode = code
            mock_subproc.return_value.stdout = stdout_val
            mock_subproc.return_value.stderr = ""
            action_id = f"rapid-toggle-{idx}"
            msg = {
                "protocol": "fleet-batch-v1",
                "action": "CONTROL_TAILSCALE",
                "action_id": action_id,
                "mode": mode,
                "target_device_ids": [self.device_id],
            }
            handle_incoming_batch_action(
                msg, self.device_id, self.report_url, self.secret, state, self.state_path, self.links_path
            )
            self.assertEqual(mock_ack.call_args.kwargs["status"], exp_status)
            self.assertEqual(mock_ack.call_args.kwargs["details"], exp_details)

        self.assertEqual(len(state["tailscale_action_results"]), len(modes_and_outputs))


class TestTailscaleIPBoundaryAndLookaroundStress(unittest.TestCase):
    """Stress testing IP boundary lookarounds in Python agent."""

    def test_regex_lookaround_matching(self):
        pattern = r"(?<![0-9a-zA-Z.])\b100\.\d{1,3}\.\d{1,3}\.\d{1,3}\b(?![0-9a-zA-Z./:])"

        # Cases that MUST match and extract valid IP
        valid_cases = [
            ("CONNECTED: 100.80.175.55", "100.80.175.55"),
            ("IP 100.64.0.1 assigned", "100.64.0.1"),
            ("100.0.0.0", "100.0.0.0"),
            ("100.255.255.255", "100.255.255.255"),
            ("tailscale0: 100.101.102.103", "100.101.102.103"),
            ("foo 100.1.2.3 bar", "100.1.2.3"),
            ("[100.1.2.3]", "100.1.2.3"),
            ("(100.1.2.3)", "100.1.2.3"),
            ("100.1.2.3!", "100.1.2.3"),
        ]
        for text, expected in valid_cases:
            match = re.search(pattern, text)
            self.assertIsNotNone(match, f"Expected match in: {text!r}")
            self.assertEqual(match.group(0), expected)
            self.assertTrue(validate_tailscale_cgnat_ip(match.group(0)))

        # Cases that MUST NOT match
        invalid_cases = [
            "CONNECTED: 100.1.2.2555",       # 4-digit octet suffix
            "prefix100.1.2.3",               # letter prefix
            "1100.1.2.3",                    # digit prefix
            "100.1.2.3suffix",               # letter suffix
            "100.1.2.3.4",                   # 5 octets
            ".100.1.2.3",                    # leading dot
            "100.1.2.3:8080",                # port attached
            "100.1.2.3/24",                  # CIDR attached
            "192.168.1.1",                   # Not 100.x
            "10.0.0.1",
            "127.0.0.1",
            "TRIGGERED",
            "",
        ]
        for text in invalid_cases:
            match = re.search(pattern, text)
            if match:
                # If it matched a substring, validate_tailscale_cgnat_ip must reject or match was invalid
                self.fail(f"Adversarial string {text!r} unexpectedly matched {match.group(0)!r}")


class TestR4ZeroRealUgPhoneCommandsAudit(unittest.TestCase):
    """Static and structural audit to guarantee zero real device interaction."""

    def test_no_adb_connect_or_live_device_commands(self):
        """Ensure no live adb connect or direct socket connections to UgPhone exist."""
        forbidden = [
            "adb connect",
            "adb -s",
            "fastboot",
            "ugphone.net",
            "ugphone.com",
        ]
        root_dir = pathlib.Path(PROJECT_ROOT)
        for path in root_dir.glob("**/*"):
            if path.is_file() and path.suffix in [".py", ".js", ".mjs", ".sh"]:
                # skip git or agents metadata or this test file itself
                if ".git" in path.parts or ".agents" in path.parts or path.name == "test_adversarial_coverage_challenger2.py":
                    continue
                content = path.read_text(encoding="utf-8", errors="ignore")
                for term in forbidden:
                    self.assertNotIn(
                        term, content,
                        f"Forbidden real device command '{term}' found in {path}!"
                    )

    def test_build_tailscale_command_uses_user0_and_suppression(self):
        """Ensure commands generated for Android strictly use --user 0 and redirection."""
        for mode in ["on", "off"]:
            cmd = build_tailscale_command(mode)
            self.assertIn("--user 0", cmd)
            self.assertIn(">/dev/null 2>&1 || true", cmd)


if __name__ == "__main__":
    unittest.main()
